pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -555,10 +579,24 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		return changes
 	}
 
+	//check sidecars status
+	sidecarStatus := status.GetSidecarsStatus(pod)
+
 	// Number of running containers to keep.
 	keepCount := 0
 	// check the status of containers.
 	for idx, container := range pod.Spec.Containers {
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
+			//if the sidecars are still becoming ready during the startup of the pod, skip any non-sidecars
+			if sidecarStatus.SidecarsPresent {
+				if sidecarStatus.ContainersWaiting && !sidecarStatus.SidecarsReady {
```

Comment:
if the sidecar is existed, whatever return code we have here, pod progress will halt.

we might need to do the followings:
1. if pod is always restart, sidecar exited, we should proceed to restart sidecar
2. if pod is restart on failure, if sidecar exited with non-zero, we should proceed to restart sidecar, if sidecar exited with zero, we **should proceed** with non-sidecar
3. if pod is "never" restart, as long as there are some sidecar running, we should proceed to start main containers.

my general feeling is that, "start non-sidecar only after all sidecars are running and ready" is not a practical statement and can cause pod to stuck somewhere

Reply:
> if pod is always restart, sidecar exited, we should proceed to restart sidecar

That's what currently happens
>pod is restart on failure, if sidecar exited with non-zero, we should proceed to restart sidecar

Also what currently happens

>if sidecar exited with zero, we should proceed with non-sidecar

This feels weird, the whole point of this feature is to give people a way to ensure that the sidecars are running before their main container, so we would lose that guarantee by doing this

> if pod is "never" restart, as long as there are some sidecar running, we should proceed to start main containers.

Same as my point above, we've lost the guarantee by doing this

Reply:
So given the case where restart policy is never and some sidecar started but existed, or the case where restart policy is on failure and some sidecar existed with 0. What do you thing would be a better way of handing the fact that pod stuck in its progress?

Maybe we can mark the pod as failed, as we will never get a chance to proceed with non-sidecars.

Reply:
I think marking the pod as failed makes sense, the sidecars permanently exited and therefore your non sidecars didn't get started so pod == failed. Similar to what happens if your pod's init containers fail with restart policy == never